README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

[![Build
Status](https://travis-ci.org/shuowencs/SortedEffects.svg?branch=master)](https://travis-ci.org/shuowencs/SortedEffects)

# SortedEffects

The R package **SortedEffects** implements econometric methods in
Chernozhukov, Fernandez-Val and Luo (2018) and reports *sorted partial
effects*, a collection of estimated partial effects sorted in increasing
order and indexed by percentiles. In contrast to the classical average
partial effect, sorted effect curves completely represent range of
heterogeneous partial effects, which are common in nonlinear predictive
models. The package further allows users to conduct classification
analysis with sorted effects and visualize the results. You can install
the current version of with:

``` r
install.packages("devtools")
devtools::install_github("yuqimemeda/SortedEffects")
```

This repository hosts the (commented) source code and data. For more
theoretical details, please refer to the [original
paper](https://www.econometricsociety.org/publications/econometrica/2018/11/01/sorted-effects-method-discovering-heterogeneous-effects-beyond)
.

This document is structured to answer three curiosity-sparked questions:

  - What do you mean by (heterogeneous) partial effects and
    classification analysis?
  - What function commands does **SortedEffects** offer?
  - How do I use **SortedEffects** for my research?

I will use racial-based discrimination in mortgage lending as an
application to illustrate the basic functionality of **SortedEffects**.
For more detailed explanations of function commands, refer to the
companion vignette.

## Econometrics Background

Many economic inquiries reduce to the following question: holding all
else (control variables) fixed, how does change in a key variable affect
the outcome? Depending on the assumptions we impose, such effects have
different names like treatment or marginal effects. To ease conversation
we call them **partial effects**. To answer the question, economists
often work with nonlinear models. In particular, consider the following
two kinds of nonlinearity: nonlinearity in key variables of interest or
in parameters. The first kind includes OLS and quantile regressions in
which the variable of interest is interacted with control variables. The
second kind includes generalized linear models like Logit and Probit. It
turns out that the partial effects derived under nonlinear settings vary
among obervational units. We call these different effects as
**heterogenous partial effects**.

Before jumping to a concrete example, let’s fix some notations. Let
\(Y\) denote outcome variable of interest and \(X=(T, W)\) denote
regressors. In particular, \(T\) is the variable in interest while \(W\)
means control variables. We link between \(Y\) and \(X\) with a
predictive function \(g(X)\). With more assumptions \(g\) can represent
causal relationship, but it’s not necessary for our purposes.

The economic question is: holding \(W\) fixed, how do changes in \(T\)
affect \(Y\)? The mathematical expression for partial effects is
\[\Delta(x) = g(t_{1}, w)-g(t_{0}, w)\]

for discrete \(T\) and \[\Delta(x) = \partial_{t}g(t, w)\] for \(T\)
being continuous. It turns out that if we work with nonlinear models,
\(\Delta(x)\) depends on values of \(w\).

Here is a concrete probit example:
\(Y=\Phi(T \beta_{1}+W \beta')+\epsilon\), where \(\Phi\) denotes CDF of
standard normal distribution. If \(T\) is continuous, then the partial
effect is \[\Delta(x)=\phi(T\beta_{1}+W\beta')\beta_{1}.\] It is obvious
that \(\Delta(x)\) depends on values of \(W\). Since sample observations
will have different values of \(W\), \(\Delta(x)\) features
heterogeneity. Because \(\Delta(x)\) is unobservable in data, we use
econometric techniques to get an estimate. However, no estimation is
accurate. Therefore we need to quantify the uncertainty of estimation,
and in econometrics this is called **inference**.

It turns out hard to do inference on the estimate of partial effects, so
in applied economic work researchers often report the average partial
effect as an alternative (averaged over the empirical distribution of
\(x\)). But by definition such reports neglect heterogeneity.

Chernozhukov, Fernandez-Val and Luo (forthcoming, *Econometrica*)
provide theoretical underpinnings to conduct estimation and inference on
the partial effect values at any quantile in interest. For example, if
we want to know the partial effects from 2% to 98%, then the method will
report estimate of all these effects and quantification of estimation
uncertainty (confidence bands). Because the estimated partial effects
are sorted in an increasing order, we call them **sorted partial effect
(SPE)**.

With SPE, we can classify observational units into groups based on how
greatly they are affected. For example, researchers might be curious
whether mostly affected observational units have different
characteristics compared to peers who are least affected.

## Example 1: Racial-Based Discrimination in Mortgage Lending

We use data on mortgage applications in Boston from 1990. The Federal
Reserve Bank of Boston collected these data in relation to the Home
Mortgage Disclosure Act (HMDA), which was passed to monitor minority
access to the mortgage market. Providing better access to credit markets
can arguably help the disadvantaged groups escape poverty traps. To
access the data, type in the following command

``` r
data("mortgage")
```

The sample includes 2380 observations corresponding to 2041 white
applicants and 339 black applicants.

We estimate a binary response model where the outcome variable \(Y\) is
an indicator for mortgage denial, the key covariate \(T\) is an
indicator for the applicant being black, and the controls \(W\) contain
financial and other characteristics of the applicant that banks take
into account in the mortgage decisions. The regression formula is
specified as

``` r
fm <- deny ~ black + p_irat + hse_inc + ccred + mcred + pubrec + ltv_med + ltv_high + denpmi + selfemp + 
      single + hischl
```

We use `SPE` command to estimate SPE and APE.

``` r
test <- spe(fm = fm, data = mortgage, var = "black", method = "logit",
            us = c(2:98)/100, b = 500, bc = TRUE)
```

In this line of code, `var.T = "black"` tells the package that \(T\) is
`black`, while `us` specifies the SPE quantile index to range from 2% to
98%. `B` refers to the number of bootstrap for inference. For accuracy
we recommend setting `B` to be 500. The result `ex` is a list containing
SPE and APE with corresponding confidence bands, and for visualization
we can use the `plot.SPE` command as follows:

``` r
plot(x = test, ylim = c(0, 0.25), ylab = "Change in Probability",
     main = "APE and SPE of Being Black on the Prob of Mortgage Denial",
     sub = "Logit Model")
```

We see that the SPE ranges from 2% to 14% while APE is 6%. It’s clear
that reporting APE masks rich heterogeneity of the partial effects.

Now that we know being black affects mortgage denial probability
differently among applicants, we can ask a follow-up question: what are
the average characteristics of most and least affected applicants? Such
questions are referred to as **classification analysis**: first classify
observational units into most/least affected groups and then study
objects of interest like mean and distributions.

To conduct estimation and inference on classification analysis, we use
the `u.CA` command. We first choose the characteristics in interest:
Deny, Black, Debt-to-income, Expenses-to-income, Bad consumer credit,
Bad mortgage credit, Credit problems, Denied mortgage insurance, Medium
loan-to-house, High loan-to-house, Self employed, Single and High school
grad. The corresponding specification is

``` r
t <- c("deny", "p_irat", "black", "hse_inc", "ccred", "mcred", "pubrec",
       "denpmi", "selfemp", "single", "hischl", "ltv_med", "ltv_high")
```

Then we issue to `u.CA` command:

``` r
CA <- ca(fm = fm, data = mortgage, var = "black", method = "logit",
         cl = "both", t = t, b = 500, bc = TRUE)
```

The specification `cl=matrix(c(1,0,0,1), nrow=2)` means we tabulate the
mean of the characteristics for two groups. The result `CA` is a list
containing the estimate, standard errors and joint p-values. We can
tabulate the result as follows:

``` r
summary(CA)
```

We can see that the characteristics are different for these two groups.


NEWS.md
NEWS
================
Shuowen Chen
9/6/2021

# SortedEffects 1.6.0.
Changes from SortedEffects 1.5.0. The previous version imports a package rlist. Upon receving a warning from CRAN about reverse dependencies, we revise the code so that the package no longer relies on rlist. 

Changes from SortedEffects 1.4.0. Renormalize the weights for stable estimates using glm.  

Changes from SortedEffects 1.3.0. For nonparametric bootstrap, we use the multinomial weight resampling to produce more stable bootstrap estimates. 

Changes from SortedEffects 1.2.0. The spe function in the 1.2.0 version produces far wider nonparametric bootstrap confidence bands for some specifications. We fix that issue. 

Changes from SortedEffects 1.1.0. The 1.1.0 version has a check problem due to a recent upgrade of package 'tibble' to version 3.0. The package doesn't really need tibble package, so we change the data type. 


R/data.R
#' Mortgage Denial
#'
#' @format Contains the data on mortgage application in Boston from 1990,
#' (Munnell et al., 1996.)
#' We obtain the data from the companion website of Stock and Watson (2011).
#' The file contains the following variables:
#' \describe{
#' \item{deny}{indicator for mortgage application denied}
#' \item{p_irat}{monthly debt to income ratio}
#' \item{black}{indicator for black applicant}
#' \item{hse_inc}{monthly housing expenses to income ratio}
#' \item{loan_val}{loan to assessed property value ratio}
#' \item{ccred}{consumer credit score with 6 categories. 1 if no "slow"
#' payments or delinquencies, 2 if one or two "slow" payments or delinquencies,
#' 3 if more than two "slow" payments or delinguencies, 4 if insufficient
#' credit history for determination, 5 if delinquent credit history with
#' payment 60 days overdue, and 6 if delinquent credit history with payments
#' 90 days overdue.}
#' \item{mcred}{mortgage credit score with 4 categories. 1 if no late mortgage
#' payments, 2 if no mortgage payment history, 3 if one or two late mortage
#' payments, and 4 if more than two late mortgages payments}
#' \item{pubrec}{indicator for any public record of credit problems: bankruptcy
#' , charge-offs, collection actions}
#' \item{denpmi}{indicator for applicant applied for mortgage insurance and was
#' denied}
#' \item{selfemp}{indicator for self-employed applicant}
#' \item{single}{indicator for single applicant}
#' \item{hischl}{indicator for high school graduated applicant}
#' \item{probunmp}{1989 Massachusetts unemployment rate in the applicant's
#' history}
#' \item{condo}{indicator for unit is a condominium}
#' \item{ltv_med}{indicator for medium loan to property value ratio [.80, .95]}
#' \item{ltv_high}{indicator for high loan to property value ratio >.95}
#'}
#'@source Munnell, Alicia, Geoffrey Tootell, Lynn Browne, and James McEneaney,
#'"Mortgage Lending in Boston: Interpreting HMDA Data", The American Economic
#'Review, 1996.
"mortgage"

#' Wage Data
#'
#' @format Consists of white, non-hispanic individuals aging from 25 to 64 and
#' working more than 35 hours per week during at least 50 weeks of the year.
#' Excludes self-employed, individuals living in group quarters; individuals in
#' the military, agricultural or private household sectors; individuals with
#' inconsistent reports on earnings and employment status; individuals with
#' allocated or missing information in any of the variables used in the
#' analysis; individuals with hourly wage rate below $3. Contains 32,523
#' workers including 18,137 men and 14,382 women. The file contains the
#' following variables:
#' \describe{
#' \item{lnw}{log of hourly wages}
#' \item{weight}{CPS sampling weight}
#' \item{female}{gender indicator: 1 if female}
#' \item{exp1}{max(age-years of educ-7, 0)}
#' \item{exp2}{exp1^2/100}
#' \item{exp3}{exp1^3/100}
#' \item{exp4}{exp1^4/100}
#' \item{occ}{Aggregated occupation with 5 categories: managers, service,
#' sales, construction and production.}
#' \item{ind}{Aggregated industry with 12 categories: minery, construction,
#' manufacture, retail, transport, information, finance, professional,
#' education, leisure, services, public.}
#' \item{educ}{Education attainment with 5 categories: lhs (less than high
#' school graduate, years of educ < 12), hsg (high school graduate: years of
#' educ = 12), sc (some college: 13<=years of educ<=15), cg (college:
#' 16<=years of educ<=17), ad (advanced degree: years of educ>=18).}
#' \item{ms}{Marital Status with 5 categories: married, widowed, separated,
#' divorced, and nevermarried.}
#' \item{region}{Regions with 4 categories: mw (midwest), so (south),
#' we (west), ne (northeast).}
#' }
#' @source U.S. March Supplement of the Current Population Survey (CPS) in 2015.
#'
"wage2015"


R/subpop.R
#' Inference on Most and Least Affected Groups
#'
#' \code{subpop} conducts set inference on the groups of most and least
#' affected. When \code{subgroup = NULL}, output is for whole sample. Otherwise
#' the results are subgroup. The output of \code{subpop} is a list
#' containing six components: \code{cs_most}, \code{cs_least}, \code{u},
#' \code{subgroup}, \code{most} and \code{least}. As the names
#' indicate, \code{cs_most} and \code{cs_least} denote the confidence sets for
#' the most and least affected units. \code{u} stores the u-th most and least
#' affected index. \code{subgroup} stores the indicators for subpopulations.
#' \code{most} and \code{least} store the data of the most and
#' least affected groups. The confidence sets can be visualized using the
#' \code{\link{plot.subpop}} command while the two groups can be tabulated via
#' the \code{\link{summary.subpop}} command.
#'
#' @param fm          Regression formula
#' @param data        The data in use
#' @param method      Models to be used for estimating partial effects. Four
#'                    options: \code{"logit"} (binary response),
#'                    \code{"probit"} (binary response), \code{"ols"}
#'                    (interactive linear with additive errors), \code{"QR"}
#'                    (linear model with non-additive errors). Default is
#'                    \code{"ols"}.
#' @param var_type    The type of parameter in interest. Three options:
#'                    \code{"binary"}, \code{"categorical"},
#'                    \code{"continuous"}. Default is \code{"binary"}.
#' @param var         Variable T in interset. Should be a character.
#' @param compare     If parameter in interest is categorical, then user needs
#'                    to specify which two category to compare with. Should be
#'                    a 1 by 2 character vector. For example, if the two levels
#'                    to compare with is 1 and 3, then \code{c=("1", "3")},
#'                    which will calculate partial effect from 1 to 3. To use
#'                    this option, users first need to specify \code{var} as a
#'                    factor variable.
#' @param subgroup    Subgroup in interest. Default is \code{NULL}.
#'                    Specifcation should be a logical variable. For example,
#'                    suppose data contains indicator variable for women
#'                    (female if 1, male if 0). If users are interested in
#'                    women SPE, then users should specify
#'                    \code{subgroup = data[, "female"] == 1}.
#' @param samp_weight Sampling weight of data. Input should be a n by 1 vector,
#'                    where n denotes sample size. Default is \code{NULL}.
#' @param taus        Indexes for quantile regression.
#'                    Default is \code{c(5:95)/100}.
#' @param u           Percentile of most and least affected. Default is set to
#'                    be 0.1.
#' @param alpha       Size for confidence interval. Shoule be between 0 and 1.
#'                    Default is 0.1
#' @param b           Number of bootstrap draws. Default is set to be 500.
#' @param parallel    Whether the user wants to use parallel computation.
#'                    The default is \code{FALSE} and only 1 CPU will be used.
#'                    The other option is \code{TRUE}, and user can specify
#'                    the number of CPUs in the \code{ncores} option.
#' @param ncores      Number of cores for computation. Default is set to be
#'                    \code{detectCores()}, which is a function from package
#'                    \code{parallel} that detects the number of CPUs on the
#'                    current host. For large dataset, parallel computing is
#'                    highly recommended since bootstrap is time-consuming.
#' @param seed        Pseudo-number generation for reproduction. Default is 1.
#' @param boot_type   Type of bootstrap. Default is \code{"nonpar"}, and the
#'                    package implements nonparametric bootstrap. The
#'                    alternative is \code{"weighted"}, and the package
#'                    implements weighted bootstrap.
#'
#' @examples
#' data("mortgage")
#' ### Regression Specification
#' fm <- deny ~ black + p_irat + hse_inc + ccred + mcred + pubrec +
#'    ltv_med + ltv_high + denpmi + selfemp + single + hischl
#' ### Issue the subpop command
#' set_b <- subpop(fm, data = mortgage, method = "logit", var = "black",
#' u = 0.1, alpha = 0.1, b = 50)
#'
#' @importFrom Hmisc wtd.quantile
#' @importFrom boot boot
#' @importFrom stats quantile rexp qnorm
#' @importFrom parallel detectCores
#' @importFrom pbapply setpb startpb closepb
#' @export
subpop <- function(fm, data, method = c("ols", "logit", "probit", "QR"),
                   var_type = c("binary", "continuous", "categorical"),
                   var, compare, subgroup = NULL, samp_weight = NULL,
                   taus = c(5:95)/100, u = 0.1, alpha = 0.1, b = 500, seed = 1,
                   parallel = FALSE, ncores = detectCores(),
                   boot_type = c("nonpar", "weighted")) {
  #---------------------- Stopping Conditions ---------------------------
  if (alpha >= 1 || alpha <= 0) stop("Please specify a correct size for
                                     hypothesis testing between 0 and 1.")
  if (u >= 1 || u <= 0) stop("Please provide a group classification
                             quantile between 0 and 1.")
  # ------ Replace Null samp_weight specification
  if (is.null(samp_weight)) samp_weight <- rep(1, dim(data)[1])
  samp_weight <- samp_weight/mean(samp_weight) # renormalize
  # ------ Matching Arguments
  method <- match.arg(method)
  var_type <- match.arg(var_type)
  boot_type <- match.arg(boot_type)
  # ------ 1. Call to estimate PE and PEsub
  output <- suppressWarnings(peestimate(fm, data, samp_weight, var_type, var, compare, method,
                       subgroup, taus))
  pe_est <- output$pe_est
  # ----- 2. u-most and least Affected Groups
  if (method != "QR") {
    if (is.null(subgroup)) {
      # Threshold Values for u-most/least Affected for WHOLE sample
      effect_high <- wtd.quantile(pe_est, samp_weight, 1 - u)
      effect_low <- wtd.quantile(pe_est, samp_weight, u)
      # Get weight for the two groups. If weight is NULL, it does nothing
      # Two affected groups
      high_affected <- data[pe_est >= effect_high, ]
      low_affected <- data[pe_est <= effect_low, ]
    } else {
      # SUB sample
      pesub_est <- output$pesub_est
      pesub_w <- output$samp_weight_sub
      # Threshold Values for u-most/least Affected
      effect_sub_high <- wtd.quantile(pesub_est, pesub_w, 1 - u)
      effect_sub_low <- wtd.quantile(pesub_est, pesub_w, u)
      subdata <- data[subgroup, ]
      high_affected <- subdata[pesub_est >= effect_sub_high, ]
      low_affected <- subdata[pesub_est <= effect_sub_low, ]
    }
  } # QR requires special treatment due to stacking of quantile indices
  if (method == "QR") {
    if (is.null(subgroup)) {
      # This is where QR is different.
      effect_high <- wtd.quantile(pe_est, matrix(samp_weight, ncol = 1,
                                                 nrow = nrow(pe_est),
                                                 byrow = FALSE), 1 - u)
      effect_low <- wtd.quantile(pe_est, matrix(samp_weight, ncol = 1,
                                                nrow = nrow(pe_est),
                                                byrow = FALSE), u)
      # Kronecker function doesn't work for all data types, so I wrote this
      # alternative
      mesh <- data[rep(1:nrow(data), times = length(taus)), ]
      # Two Affected Groups
      high_affected <- mesh[pe_est >= effect_high, ]
      low_affected <- mesh[pe_est <= effect_low, ]
    } else {# Subsample: use PEsub_est
      pesub_est <- output$pesub_est
      pesub_w <- output$samp_weight_sub
      # Threshold Values for u-most/least Affected
      effect_sub_high <- wtd.quantile(pesub_est, pesub_w, 1 - u)
      effect_sub_low <- wtd.quantile(pesub_est, pesub_w, u)
      subdata <- data[subgroup, ]
      mesh <- subdata[rep(1:nrow(subdata), times = length(taus)), ]
      # Two Affected Groups
      high_affected <- mesh[pesub_est >= effect_sub_high, ]
      low_affected <- mesh[pesub_est <= effect_sub_low, ]
    }
  }
  # ---------------------------- 3. Bootstrap Samples -------------------------
  # set a bootstrap counting variable for the purpose of showing a progress bar
  rep_count <- 1
  # Resampling function for weighted bootstrap
  data_rg <- function(data, mle){
    n <- dim(data)[1]
    # Exponential weights
    multipliers  <- rexp(n)
    # Sampling weight of data
    weight <- samp_weight * multipliers / sum(multipliers) * 20000
    data$.w <- weight
    return(data)
  }
  # Resampling function for nonparametric bootstrap
  data_non <- function(data, mle){
    n <- dim(data)[1]
    multipliers <- as.vector(table(factor(sample(n,n,replace = T),
                                          levels = c(1:n))))
    # Sampling weight of data.bs
    weight <- (multipliers/sum(multipliers)) * samp_weight * 20000
    data$.w <- weight
    return(data)
  }
  # Function that computes bootstrap statistics in each draw
  boot_stat_weight <- function(data){
    # set up a progress bar to document the bootstrap progress
    setpb(pb, rep_count)
    rep_count <<- rep_count + 1
    out_bs <- suppressWarnings(peestimate(fm, data, samp_weight = data$.w, var_type, var,
                         compare, method, subgroup, taus))
    pe_est_bs <- out_bs$pe_est
    if (method == "QR") {
      effect_high_bs <- wtd.quantile(pe_est_bs, matrix(data$.w, ncol = 1,
                                                       nrow = nrow(pe_est),
                                                       byrow = FALSE), 1 - u)
      effect_low_bs <- wtd.quantile(pe_est_bs, matrix(data$.w, ncol = 1,
                                                      nrow = nrow(pe_est),
                                                      byrow = FALSE), u)
    } else {
      effect_high_bs <- wtd.quantile(pe_est_bs, data$.w, 1 - u)
      effect_low_bs <- wtd.quantile(pe_est_bs, data$.w, u)
    }
    if (!is.null(subgroup)) {
      pesub_est_bs <- out_bs$pesub_est
      pesub_w_bs <- out_bs$samp_weight_sub
      effect_sub_high_bs <- wtd.quantile(pesub_est_bs, pesub_w_bs, 1 - u)
      effect_sub_low_bs <- wtd.quantile(pesub_est_bs, pesub_w_bs, u)
      return(c(effect_sub_high_bs, effect_sub_low_bs, pesub_est_bs))
    } else {
      return(c(effect_high_bs, effect_low_bs, pe_est_bs))
    }
  }
  # ----- 4. Conduct Bootstrap
  set.seed(seed)
  if (parallel == FALSE) ncores <- 1
  if (boot_type == "nonpar") {
    data$.w <- samp_weight
    cat(paste("Using", ncores, "CPUs now.\n"))
    pb <- startpb(min = 0, max = b)
    result_boot <- boot(data = data, statistic = boot_stat_weight,
                        sim = "parametric", ran.gen = data_non, mle = 0,
                        parallel = "multicore", ncpus = ncores, R = b)
    closepb(pb)
    data$.w <- NULL
  } else if (boot_type == "weighted") {
    data$.w <- samp_weight
    cat(paste("Using", ncores, "CPUs now.\n"))
    pb <- startpb(min = 0, max = b)
    result_boot <- boot(data = data, statistic = boot_stat_weight,
                        sim = "parametric", ran.gen = data_rg, mle = 0,
                        parallel = "multicore", ncpus = ncores, R = b)
    closepb(pb)
    data$.w <- NULL
  }

  # -----5. Analysis for the subpopulation -=
  if (is.null(subgroup)) {
    # (a) PE
    # MOST Affected Group confidence set
    is_he <- submost(pe_est, result_boot$t[, 3:(length(pe_est) + 2)],
                     result_boot$t[, 1], effect_high, alpha, b)
    # LEAST Affected Group confidence set
    is_le <- subleast(pe_est, result_boot$t[, 3:(length(pe_est) + 2)],
                      result_boot$t[, 2], effect_low, alpha, b)
    output <- list(cs_most = is_he, cs_least = is_le, u = u,
                   subgroup = subgroup, most = high_affected,
                   least = low_affected)
  } else{
    # (b) PEsub
    is_he_sub <- submost(pesub_est, result_boot$t[, 3:(length(pesub_est) + 2)],
                         result_boot$t[, 1], effect_sub_high, alpha, b)
    is_le_sub <- subleast(pesub_est, result_boot$t[, 3:(length(pesub_est) + 2)],
                          result_boot$t[, 2], effect_sub_low, alpha, b)
    output <- list(cs_most = is_he_sub, cs_least = is_le_sub, u = u,
                   subgroup = subgroup, most = high_affected,
                   least = low_affected)
  }
  # claim output as a class
  output <- structure(output, class = "subpop")
  return(output)
}

# -----Two Auxiliary Functions
# Implementing algorithm: Output is confidence set indicator
# Most affected group
submost <- function(est_pe, bs_pe, bs_u, est_u, alpha, b){
  # Find the min (to implement the sup condition as in the paper)
  is_0 <- rank(abs(est_pe - est_u)) == min(rank(abs(est_pe - est_u)))
  draws <- bs_pe - matrix(bs_u, nrow = b, ncol = length(est_pe)) -
    matrix(est_pe - est_u, nrow = b, ncol = length(est_pe), byrow = TRUE)
  bse <- (apply(draws, 2, quantile, .75, na.rm = TRUE) -
            apply(draws, 2, quantile, .25, na.rm = TRUE))/(qnorm(0.75) - qnorm(.25))
  #bm <- apply(draws, 2, mean) # bias estimator
  zs <- apply(-draws[, is_0] / matrix(bse[is_0], nrow = b,
                                      ncol = length(bse[is_0]), byrow = TRUE),
              1, max, na.rm = TRUE)
  crt <- quantile(zs, 1 - alpha)  #critical value
  is_he <- -(est_pe - est_u) / bse <= crt
  return(is_he)
}
# Least affected group
subleast <- function(est_pe, bs_pe, bs_u, est_u, alpha, b) {
  # Find the min (to implement the sup condition as in the paper)
  is_0 <- rank(abs(est_pe - est_u)) == min(rank(abs(est_pe - est_u)))
  draws <- bs_pe - matrix(bs_u, nrow = b, ncol = length(est_pe)) -
    matrix(est_pe - est_u, nrow = b, ncol = length(est_pe), byrow = TRUE)
  bse <- (apply(draws, 2, quantile, .75, na.rm = TRUE) -
            apply(draws, 2, quantile, .25, na.rm = TRUE))/(qnorm(0.75) - qnorm(.25))
  #bm <- apply(draws, 2, mean) # bias estimator
  zs <- apply(draws[, is_0] / matrix(bse[is_0], nrow = b,
                                     ncol = length(bse[is_0]), byrow = TRUE),
              1, max, na.rm = TRUE)
  crt <- quantile(zs, 1 - alpha)  #critical value
  is_le <- (est_pe - est_u) / bse <= crt
  return(is_le)
}

# ----- Summary of subpop output
#' Return the output of \code{\link{subpop}} function.
#'
#' The \code{\link{subpop}} function stores the most and least affected groups.
#' This command allows users to see these two groups and their corresponding
#' characteristics. The command also allows users to check the summary
#' statistics of variables in interest, which can be useful for plotting the
#' projections plot via the \code{\link{plot.subpop}} method.
#'
#' @param object    Output of \code{\link{subpop}} command.
#' @param vars      The variables that users want to see the summary
#'                  statistics. The default is \code{NULL} and the command
#'                  shows all variables. The summary statistcs include min
#'                  1st quartile, median, mean, 3rd quartile and the max.
#' @param ...       additional arguments affecting the summary produced.
#'
#' @examples
#' data("mortgage")
#' ### Regression Specification
#' fm <- deny ~ black + p_irat + hse_inc + ccred + mcred + pubrec +
#'    ltv_med + ltv_high + denpmi + selfemp + single + hischl
#' ### Issue the subpop command
#' set_b <- subpop(fm, data = mortgage, method = "logit", var = "black",
#' u = 0.1, alpha = 0.1, b = 50)
#' ### Produce summary of two variables
#' groups <- summary(set_b, vars = c("p_irat", "hse_inc"))
#' @export
summary.subpop <- function(object, vars = NULL, ...) {
  # Retrieve the two affected groups from the subpop object
  high_affected <- object$most
  low_affected <- object$least
  # report all variables if users don't specify a particular interest
  if (is.null(vars)) vars <- colnames(high_affected)
  # an axiliary function that reports summary statistics
  sumstats <- function(x) {
    min <- min(x, na.rm = TRUE)
    max <- max(x, na.rm = TRUE)
    median <- median(x, na.rm = TRUE)
    mean <- mean(x, na.rm = TRUE)
    first_quartile <- quantile(x, 0.25, na.rm = TRUE)
    third_quartile <- quantile(x, 0.75, na.rm = TRUE)
    output <- matrix(rbind(min, first_quartile, median, mean, third_quartile, max))
    #rownames(output) <- c("Min", "1st Qu", "Median", "Mean", "3rd Qu", "Max")
    return(output)
  }
  info_m <- matrix(as.numeric(unlist(high_affected[,vars])), ncol = length(vars))
  info_l <- matrix(as.numeric(unlist(low_affected[,vars])), ncol = length(vars))
  info_m <- apply(info_m, 2, sumstats)
  info_l <- apply(info_l, 2, sumstats)
  rownames(info_m) <- rownames(info_l) <- c("Min", "1st Quartile", "Median",
                                            "Mean", "3rd Quartile", "Max")
  colnames(info_m) <- colnames(info_l) <- vars
  output <- list(most_affected = high_affected, least_affected = low_affected,
                 stats_most = info_m, stats_least = info_l)
  return(output)
}

# ----- Plotting (2-dimensional projection plots of two specified variables)
#'
#' Plot 2-dimensional projections of variables in interest.
#'
#' Takes output from \code{\link{subpop}} command as inputs and plots
#' 2-dimensional projection plots of two specified variables. If a
#' variable in interest is of type factor, then the user must put it on
#' the y-axis. If the variable on the y-coordinate is a factor, range of
#' y-axis is set to be the factor level. Otherwise, users can use
#' \code{\link{summary.subpop}} to know the ranges of variables in the
#' two groups.
#'
#' @param x           Output of \code{\link{subpop}} command.
#' @param varx        The name of the variable to be plotted on the x-axis.
#' @param vary        The name of the variable name to be plotted on the
#'                    y-axis.
#' @param xlim        The range of x-axis. Default is \code{NULL}.
#' @param ylim        The range of y-axis. Default is \code{NULL}. If the
#'                    variable on the y-coordinate is a factor, the default
#'                    will set it to be the factor level, and users don't
#'                    need to specify \code{ylim}.
#' @param main        Main title of the plot. Default is \code{NULL}.
#' @param sub         Sub title of the plot. Default is NULL.
#' @param xlab        x-axis label. Default is \code{NULL}.
#' @param ylab        y-axis label. Default is \code{NULL}.
#' @param overlap     Whether user wants to allow observations included in both
#'                    confidence sets. Default is \code{FALSE}, and the plot
#'                    drops the overlapped observations.
#' @param ...         Graphics parameters to be passed to the plotting
#'                    routines.
#'
#' @examples
#' data("mortgage")
#' ### Regression Specification
#' fm <- deny ~ black + p_irat + hse_inc + ccred + mcred + pubrec +
#'    ltv_med + ltv_high + denpmi + selfemp + single + hischl
#' ### Issue the subpop command
#' set_b <- subpop(fm, data = mortgage, method = "logit", var = "black",
#' u = 0.1, alpha = 0.1, b = 50)
#' ### Plotting
#' plot(set_b, varx = mortgage$p_irat, vary = mortgage$hse_inc,
#'      xlim = c(0, 1.5), ylim = c(0, 1.5), xlab = "Debt/Income",
#'      ylab = "Housing expenses/Income", overlap = TRUE)
#' @importFrom graphics plot polygon lines legend
#' @export
plot.subpop <- function(x, varx, vary, xlim = NULL, ylim = NULL, main = NULL,
                        sub = NULL, xlab = NULL, ylab = NULL, overlap = FALSE,
                        ...) {
  if (is.factor(varx)) stop("Variable on the x-axis cannot be a factor.")
  is_he <- x$cs_most
  is_le <- x$cs_least
  subgroup <- x$subgroup
  u <- x$u
  if (!is.factor(vary)) {
    plot(varx, vary, type = "n", xlim = xlim, ylim = ylim,
         log = "", main, sub, xlab, ylab, col = 4, pch = 20, lwd = 2)
  } else {
    plot(varx, vary, type = "n", xlim = xlim, ylim = NULL,
         log = "", main, sub, xlab, ylab, col = 4, pch = 20, lwd = 2,
         yaxt = 'n')
    axis(side = 2, at = c(1:nlevels(vary)),
         labels = substr(levels(vary), 1, 3))
  }
  if (is.null(subgroup)) {
    # Full sample
    if (overlap == TRUE) {
      points(varx[is_he], vary[is_he], col = 4, pch = 20, lwd = 2)
      points(varx[is_le], vary[is_le], col = 'lightblue1', pch = 1, lwd = 2)
    } else {
      points(varx[is_he & !is_le], vary[is_he & !is_le], col = 4,
             pch = 20, lwd = 2)
      points(varx[is_le & !is_he], vary[is_le & !is_he], col = 'lightblue1',
             pch = 1, lwd = 2)
    }
  } else {
    # Sub sample
    subx <- varx[subgroup]
    suby <- vary[subgroup]
    if (overlap == TRUE) {
      points(subx[is_he], suby[is_he], col = 4, pch = 20, lwd = 2)
      points(subx[is_le], suby[is_le], col = 'lightblue1', pch = 1, lwd = 2)
    } else {
      points(subx[is_he & !is_le], suby[is_he & !is_le], col = 4,
             pch = 20, lwd = 2)
      points(subx[is_le & !is_he], suby[is_le & !is_he], col = 'lightblue1',
             pch = 1, lwd = 2)
    }
  }
  legend('topleft', c(paste0(u*100, "% Most"), paste0(u*100, "% Least")),
         col = c(4, 'lightblue1'), pch = c(20, 1), horiz = F, bty = 'n')
}
